### PR TITLE
Fix `needless_collect`'s tendency to suggest code requiring multiple mutable borrows of the same value.

### PR DESCRIPTION
**Pull request by Blckbrry-Pi**
Fixes error specified in #7975.

changelog: [`needless_collect`] no longer suggests removal of `collect` when removal would create code requiring mutably borrowing a value multiple times.

### CODE REVIEW

File: clippy_lints/src/loops/needless_collect.rs [hunk 1]
```diff
@@ -237,12 +311,61 @@ impl<'a, 'tcx> Visitor<'tcx> for UsedCountVisitor<'a, 'tcx> {
...
     visitor.visit_block(block);
-    if visitor.seen_other { None } else { Some(visitor.uses) }
+    if visitor.seen_other {
+        None
+    } else {
+        Some(visitor.uses.into_iter().flatten().collect())
+    }
+}
+
+#[allow(rustc::usage_of_ty_tykind)]
```
Comment (llogiq) [inline clippy_lints/src/loops/needless_collect.rs:338]: You can replace the `TyKind::` with `ty::` and `use rustc_middle::ty` above, then remove the `#[allow(..)]` here.

**Review by llogiq — COMMENTED** [review]
This is already looking quite good. I only have some minor readability nits.

**Comment by Blckbrry-Pi** [comment]
I have fixed the suggestions.  Thank you for making them!

**Comment by llogiq** [comment]
Thank you! @bors r+

**Pull request merged**